### DISCUSSION TITLE
previously using Paint from Shape.drawShape

### DISCUSSION THREAD

**Discussion by shrift** [patrykandpatrick/vico#848] [Questions] [answered]
### Question

Hello!

I was previously using the Paint passed from Shape's drawShape. With v2.0.0-alpha.28 this override is gone, and none of the alternatives mentioned in the release notes have an override that includes a Paint: https://patrykandpatrick.com/vico/releases/2.0.0-alpha.28.

Is that the intent, to no longer include a paint for any of these drawing object overrides?

To illustrate I used to do this:
```
val ringShape: Shape = object : Shape {
    override fun drawShape(
        context: DrawContext,
        paint: Paint,
        path: Path,
        left: Float,
        top: Float,
        right: Float,
        bottom: Float,
    ) {
        val width = right - left
        val height = bottom - top
        paint.style = Paint.Style.STROKE
        paint.strokeWidth = 5.5f
        context.canvas.drawCircle(left + (width / 2), top + (height / 2), width / 2, paint)
    }
}
```
And now I do this:
```
fun ringShape(color: Color): Component {
    return object : Component {
        override fun draw(context: DrawingContext, left: Float, top: Float, right: Float, bottom: Float) {
            val paint = Paint(Paint.ANTI_ALIAS_FLAG).apply { this.color = color.toArgb() }
            val width = right - left
            val height = bottom - top
            paint.style = Paint.Style.STROKE
            paint.strokeWidth = 5.5f
            context.canvas.drawCircle(left + (width / 2), top + (height / 2), width / 2, paint)
        }
    }
}
```

This is working and I'm not specifically looking for a change, just wondering if this intended, or maybe I'm missing something.

### Vico version(s)

v2.0.0-alpha.28

### UI framework(s)

Jetpack Compose

**Comment by Gowsky** [comment]
Hello, @shrift. This isn't entirely correct. The source of the migration difficulties is that the previous implementation was already in slight misalignment with the intended use of the API.

This is easy to fix, but we need some more information to help. Can you please show us how you're using the `Shape`? Are you passing it to a `ShapeComponent`?

**Reply by shrift** [reply]
That's right; previously I was passing the created shape to a ShapeComponent that was part of a more complex LayeredComponent, eventually creating a CartesianMarker.
```
    val foreground = shapeComponent(
        shape = ringShape,
        color = AppTheme.colors.background
    )
```
Now in version alpha.28 I'm doing nearly the same, just creating the component as indicated above and using it directly:
```
val foreground = ringShape(color = AppTheme.colors.background)
```

If it's helpful, here's the CartesianMarker that I'm using with persistent markers to highlight a tapped point.

```
@Composable
internal fun rememberSelectedMarkerAttaining(): CartesianMarker {
    val color = AppTheme.colors.goalAttaining
    val background = shapeComponent(
        shape = Shape.Pill,
        color = color
    )
    val foreground = ringShape(color = AppTheme.colors.background)
    val innerOverlaying = LayeredComponent(
        rear = background,
        front = foreground,
        padding = Dimensions(3.5f),
    )
    val selectionAlpha = shapeComponent(
        shape = Shape.Pill,
        color = color.copy(alpha = .4f)
    )
    val indicator = LayeredComponent(
        rear = selectionAlpha,
        front = innerOverlaying,
        padding = Dimensions(4f),
    )
    val guideline = LineComponent(
        color = color.toArgb(),
        thicknessDp = 1f,
        shape = DashedShape(
            shape = diamondShape,
            dashLengthDp = 4f,
            gapLengthDp = 4f
        ),
    )
    return remember {
        DefaultCartesianMarker(
            indicator = { indicator },
            label = TextComponent(),
            guideline = guideline,
            valueFormatter = { _, _ -> "" },
            indicatorSizeDp = 28f
        )
    }
}
```

**Reply by Gowsky** [reply, marked as answer]
`Shape`s were never supposed to perform `Paint` mutation — this is a responsibility of `ShapeComponent`. The purpose of `Shape` is defining a particular contour, hence the name. The countour is added to the given `Path`.

Previously, `Shape`s were additionally responsible for drawing the `Path`. They were given a `Paint` instance to facilitate this. However, they were not supposed to modify it.

Now, for simplicity, drawing the `Path` is handled by the `ShapeComponent`, so the `Paint` parameter is unneeded.

Therefore, what you should do (and should have been doing) is this:

- Use a `Shape` that defines a circular contour (adds a circle to the `Path`). The library already includes such a `Shape`: `Shape.Pill` (`CorneredShape.Pill` in newer versions).
- Give the shape a stroked appearance via `ShapeComponent`.

Here's how to do it in `2.0.0-alpha.28`:

```kt
val foreground = shapeComponent(
    shape = Shape.Pill,
    color = Color.Transparent,
    strokeColor = AppTheme.colors.background,
    /* Your existing code uses a stroke width of 5.5 px (`Paint.strokeWidth` is in px). I'm applying
    the same thickness here, but you should use dp. */
    strokeThickness = with(LocalDimensions.current) { 5.5.toDp() },
)
val innerOverlaying = LayeredComponent(
    rear = background,
    front = foreground,
    padding = Dimensions.of(3.5.dp),
)
```

**Comment by shrift** [comment]
Thanks! I'll have a look at that.

**Comment by shrift** [comment]
Doing it as you suggested with
```
val foreground = shapeComponent(
    shape = Shape.Pill,
    color = Color.Transparent,
    strokeColor = AppTheme.colors.background,
    strokeThickness = with(LocalDimensions.current) { 5.5.toDp() },
)
```
worked out. Thanks!
